File: Unit_Tests/Chapter03/Exercise19_03/Exercise19_03_Unit_Test.py
import unittest
from utilities01_py.helper_python import create_session
from utilities02_py.helper_python import extract_raw_records, parse_raw_warc
from globalp.python.packtg.helper_python_global import sample_warc_loc
from pyspark import RDD

from pyspark.sql.tests import ReusedSQLTestCase



class HeavyObject():
    def __init__(self, prefix):
        self.prefix = prefix

    def get_id(self):
        return id(self)


def map_function(record):
    # new_heavy_object = HeavyObject('MapPartition')
    # id = new_heavy_object.get_id()
    list = []
    for element in partition:
        list.append(23)
    return list

def partition_function(partition):
    # new_heavy_object = HeavyObject('MapPartition')
    # id = new_heavy_object.get_id()
    list = []
    for element in partition:
        list.append(23)
    return list


def normal_map(input_rdd):
    return input_rdd.map(lambda p: partition_function(p))

def map_partitions(input_rdd):
    return input_rdd.mapPartitions(lambda p: partition_function(p))


class Exercise18_03_Unit_Test(ReusedSQLTestCase):

    def test_map_vs_mappartitions(self):
        warc_loc = sample_warc_loc
        raw_records = extract_raw_records(warc_loc, self.spark)
        warc_records = raw_records \
            .flatMap(lambda record: parse_raw_warc(record))

        input_partitions = warc_records.getNumPartitions()
        numberOfRecords = warc_records.count()

        print('@@ Number of partitions: ' + str(input_partitions))


        ids_of_map = warc_records.map()

        ids_of_mappartitions = map_partitions(warc_records)
        print(ids_of_mappartitions.count())




if __name__ == '__main__':
    unittest.main()

File: Unit_Tests/Chapter03/Exercise19_03/Exercise19_03_Unit_Test.scala
package Chapter03.Exercise19_03

import org.apache.spark.SparkFunSuite
import org.apache.spark.sql.test.SharedSQLContext
import Utilities02.HelperScala.{extractRawRecords, parseRawWarc, sampleWarcLoc}
import Utilities02.WarcRecord
import org.apache.hadoop.io.Text
import org.apache.spark.rdd.RDD

case class HeavyObject(prefix: String) {
  println(s"$prefix new heavy object created")

  def getId(): Int = System.identityHashCode(this)
}


class Exercise19_03_Unit_Test extends SparkFunSuite with SharedSQLContext {
  val inputLocWarc = sampleWarcLoc

  test("PerRecordVsPerPartition") {
    val rawRecordsWarc: RDD[Text] = extractRawRecords(inputLocWarc)
    val warcRecords: RDD[WarcRecord] = rawRecordsWarc
      .flatMap(parseRawWarc(_))

    val inputPartitions = warcRecords.getNumPartitions
    val numberOfRecords = warcRecords.count()

    val idsOfMap = warcRecords.map(_ => {
      val newHeavyObject = HeavyObject("Map")
      newHeavyObject.getId()
    })
      .distinct()
      .collect()

    println("@" * 50)

    val idsOfMapPartition = warcRecords.mapPartitions(partition => {
      val newHeavyObject = HeavyObject("MapPartition")
      partition.map(_ => newHeavyObject.getId())
    })
      .distinct()
      .collect()

    println("@" * 50)

    println(s"@@ Number of partitions: $inputPartitions")
    println(s"@@ Number of records: $numberOfRecords")
    assert(idsOfMap.length.toDouble >= numberOfRecords.toDouble * 0.95D)
    assert(idsOfMapPartition.length == inputPartitions)

  }

}
